[WIP] Left and right dockarea

vesim987 opened the pull request:
This PR add support for vertical dock client.

TODO:
- `handle_strut_partial_change` seem's to doesn't work
- do something with dock overlapping
- refactor `render_output`
- write tests

fixes: [https://github.com/i3/i3/issues/1129](https://github.com/i3/i3/issues/1129)

vesim987 commented:
Oh, i broke commit history.

okraits commented:
I guess you need to rebase your branch against the next branch.

Airblader commented:
Thanks for taking a shot at this! I really think this would be a cool feature to have in i3. As @okraits mentioned you'll need to rebase.

Once that is sorted out, two very high-level comments (I haven't reviewed the PR in detail yet):

1. It'd be good if you could write, in text form, some explanation as to how vertical dock clients fit into the container structure of i3 now. For one it'll help us understand, but I also think we'll need to document this a bit in, say, the hacking howto, as this is a change to the fundamental container structure.
2. Please don't rename things on the IPC (I just saw a "dockarea" → "hdockarea"). This breaks compatibility which we'd like to avoid when possible. Having "dockarea" and "vdockarea" is not the ideal situation, but it's better than breaking IPC clients.

Let us know via comment when you rebased the PR!

vesim987 commented:
@Airblader I've fixed this branch. There was ~170 commits difference... idk how I did this. I've just created new branch, and applied my latest commit, and renamed it.

Airblader commented:
> add left and right position to i3bar

Please don't do this in this PR. It's a completely separate topic and one that has to be decided separately as well. I currently don't think we want a vertical mode for i3bar because we have no good answer to what the statusline should look like.

Airblader commented on src/render.c:
```diff
@@ -338,41 +343,100 @@ static void render_output(Con *con) {
 
     /* First pass: determine the height of all CT_DOCKAREAs (the sum of their
      * children) and figure out how many pixels we have left for the rest */
+    Con *left_dock = 0;
```
Please use NULL over 0

Airblader commented on src/render.c:
```diff
@@ -338,41 +343,100 @@ static void render_output(Con *con) {
+    }
+
+    
+    if(bottom_dock) {
```
Again, these blocks are a lot of repetition that should be refactored.

vesim987 commented:
@Airblader do you have idea how to rewrite this part: https://github.com/vesim987/i3/blob/vertical-dockarea/src/handlers.c#L1376-L1415

op8867555 commented:
Hi there, any progress on this? I tried this PR and found some problems:
* Incorrect bar position when using multiple monitors
* Incorrect bar position when running multiple bars
* `is_docked` is not restore when loading json layout which would crash i3 when reloading

Here is some changes to fixes problems above:
https://github.com/vesim987/i3/compare/vertical-dockarea...op8867555:vertical-dockarea

I'm new to i3, please take a look and point me out if I didn't do it right. :)

Airblader commented:
@vesim987 Are you continuing your work on this?
